src/serial/MidiSessionALSA.cc

```diff
@@ -166,13 +175,250 @@ void MidiOutALSA::serialize(Archive& /*ar*/, unsigned /*version*/)
+}
+
+void MidiInALSA::plugHelper(Connector& connector_, EmuTime::param /*time*/)
+{
+	auto& midiConnector = checked_cast<MidiInConnector&>(connector_);
+	midiConnector.setDataBits(SerialDataInterface::DATA_8); // 8 data bits
+	midiConnector.setStopBits(SerialDataInterface::STOP_1); // 1 stop bit
+	midiConnector.setParityBit(false, SerialDataInterface::EVEN); // no parity
+
+	setConnector(&midiConnector); // base class will do this in a moment,
+	                              // but thread already needs it
+	connect();
+}
+
+void MidiInALSA::unplugHelper(EmuTime::param /*time*/)
+{
+	if (connected) {
+		disconnect();
+	}
+}
+
+void MidiInALSA::connect()
+{
+	destinationPort = snd_seq_create_simple_port(
+		&seq, "MIDI in pluggable",
+		SND_SEQ_PORT_CAP_WRITE | SND_SEQ_PORT_CAP_SUBS_WRITE, SND_SEQ_PORT_TYPE_MIDI_GENERIC);
+	if (destinationPort < 0) {
+		throw PlugException(
+			"Failed to create ALSA port: ", snd_strerror(destinationPort));
+	}
+
+	int err = snd_seq_connect_from(&seq, destinationPort, srcClient, srcPort);
+	if (err) {
+		snd_seq_delete_simple_port(&seq, destinationPort);
+		throw PlugException(
+			"Failed to connect to ALSA port "
+			"(", srcClient, ':', srcPort, ")"
+			": ", snd_strerror(err));
+	}
+
+	snd_midi_event_new(MidiOutDevice::MAX_MESSAGE_SIZE, &event_parser);
+	snd_midi_event_no_status(event_parser, 1);
+
+	connected = true;
+	stop = false;
+
+	thread = std::thread([this]() { run(); });
+}
+
+void MidiInALSA::disconnect()
+{
+	stop = true;
+	thread.join();
+
+	snd_midi_event_free(event_parser);
+	snd_seq_disconnect_from(&seq, destinationPort, srcClient, srcPort);
+	snd_seq_delete_simple_port(&seq, destinationPort);
+
+	connected = false;
+}
+
+void MidiInALSA::run()
+{
+	assert(isPluggedIn());
+
+	auto npfd = snd_seq_poll_descriptors_count(&seq, POLLIN);
+	std::vector<struct pollfd> pfd(npfd);
+	snd_seq_poll_descriptors(&seq, pfd.data(), npfd, POLLIN);
+
+	while (true) {
+		if (stop)
+			break;
+		if (poll(pfd.data(), npfd, 1000) > 0) {
+			snd_seq_event_t *ev = NULL;
+
+			if (auto err = snd_seq_event_input(&seq, &ev); err < 0) {
+				std::cerr << "Error receiving MIDI event: "
+					<< snd_strerror(err) << '\n';
+				continue;
+			}
+
+			if (!ev) {
+				continue;
+			}
+
+			if (ev->type == SND_SEQ_EVENT_SYSEX) {
+				std::lock_guard<std::mutex> lock(mutex);
+
+				for (auto i : xrange(ev->data.ext.len)) {
+					queue.push_back(static_cast<uint8_t*>(ev->data.ext.ptr)[i]);
+				}
+			} else {
+				std::array<uint8_t, 12> bytes = {};
+
+				auto size = snd_midi_event_decode(event_parser, bytes.data(), bytes.size(), ev);
+				if (size < 0) {
+					std::cerr << "Error decoding MIDI event: "
+						<< snd_strerror(size) << '\n';
+					snd_seq_free_event(ev);
+					break;
+				}
+
+				std::lock_guard<std::mutex> lock(mutex);
+				for (auto i : xrange(size)) {
+					queue.push_back(bytes[i]);
+				}
+			}
+			snd_seq_free_event(ev);
+			eventDistributor.distributeEvent(
+				Event::create<MidiInALSAEvent>());
```

Comment:
This sends a `MidiInALSAEvent` for each MIDI message put in the buffer, but the signal handler method only retrieves a single byte from that buffer, while more than one byte could have been inserted. Unless I'm missing something, this would lead to multi-byte MIDI messages to be processed too late.

I checked the Windows MIDI-in code and the same pattern is used there. However, I'm not convinced that the Windows MIDI-in code is bug free; it's really old code and I don't think it has seen a lot of testing.